cpp/src/Ice/OutgoingResponse.cpp

```diff
@@ -0,0 +1,316 @@
+        }
+        catch (const UnknownException& ex)
+        {
+            exceptionId = ex.ice_id();
+            replyStatus = ReplyStatus::UnknownException;
+            errorMessage = ex.unknown;
+        }
+        catch (const LocalException& ex)
+        {
+            exceptionId = ex.ice_id();
+            replyStatus = ReplyStatus::UnknownLocalException;
+            ostringstream str;
+            str << ex;
+            if (IceUtilInternal::printStackTraces)
+            {
+                str << '\n' << ex.ice_stackTrace();
+            }
+            errorMessage = str.str();
+        }
+        catch (const Exception& ex)
+        {
+            exceptionId = ex.ice_id();
+            replyStatus = ReplyStatus::UnknownException;
+            ostringstream str;
+            str << ex;
+            if (IceUtilInternal::printStackTraces)
+            {
+                str << '\n' << ex.ice_stackTrace();
+            }
+            errorMessage = str.str();
+        }
+        catch (const std::exception& ex)
+        {
+            replyStatus = ReplyStatus::UnknownException;
+            exceptionId = ex.what();
+            ostringstream str;
+            str << "c++ exception: " << exceptionId;
+            errorMessage = str.str();
+        }
+        catch (...)
+        {
+            replyStatus = ReplyStatus::UnknownException;
+            exceptionId = "unknown";
+            errorMessage = "c++ exception: unknown";
+        }
+
+        if ((current.requestId != 0) &&
+            (replyStatus == ReplyStatus::UnknownUserException || replyStatus == ReplyStatus::UnknownLocalException ||
+             replyStatus == ReplyStatus::UnknownException))
+        {
+            ostr.write(static_cast<uint8_t>(replyStatus));
+            ostr.write(errorMessage);
+        }
+
+        return OutgoingResponse{replyStatus, std::move(exceptionId), std::move(errorMessage), ostr, current};
+    }
+} // anonymous namespace
+
+OutgoingResponse::OutgoingResponse(
+    ReplyStatus replyStatus,
+    string exceptionId,
+    string errorMessage,
+    OutputStream& outputStream,
+    const Current& current) noexcept
+    : _current(current),
+      _exceptionId(std::move(exceptionId)),
+      _errorMessage(std::move(errorMessage)),
+      _replyStatus(replyStatus)
+{
+    _outputStream.swap(outputStream);
+}
+
+OutgoingResponse::OutgoingResponse(OutgoingResponse&& other) noexcept
+    : _current(std::move(other._current)),
+      _exceptionId(std::move(other._exceptionId)),
+      _errorMessage(std::move(other._errorMessage)),
+      _replyStatus(other._replyStatus)
+{
+    _outputStream.swap(other._outputStream);
+}
+
+OutgoingResponse&
+OutgoingResponse::operator=(OutgoingResponse&& other) noexcept
+{
+    _current = std::move(other._current);
+    _replyStatus = other._replyStatus;
+    _exceptionId = std::move(other._exceptionId);
+    _errorMessage = std::move(other._errorMessage);
+    _outputStream.swap(other._outputStream);
```

Comment:
If we implemented the move assignment operator on OuputStream could we then use the `= default` implementations for some these classes?

Reply:
Yes, and it would be less error-prone. It's easy to forget to update all these functions when you add a new field to OutgoingResponse.